pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -1243,86 +1247,145 @@ public CompletableFuture<Void> close() {
     }
 
     /**
-     * Close this topic - close all producers and subscriptions associated with this topic.
-     *
+     * Close this topic - close all resources associated with this topic:
+     *   1.clients (producers and consumers) {@link #asyncCloseClients()}.
+     *   2.managed ledger {@link #asyncCloseLedger()}.
+     *   3.limiters {@link #closeLimiters()}.
      * @param closeWithoutWaitingClientDisconnect don't wait for client disconnect and forcefully close managed-ledger
      * @return Completable future indicating completion of close operation
      */
     @Override
     public CompletableFuture<Void> close(boolean closeWithoutWaitingClientDisconnect) {
-        CompletableFuture<Void> closeFuture = new CompletableFuture<>();
-
         lock.writeLock().lock();
         try {
             // closing managed-ledger waits until all producers/consumers/replicators get closed. Sometimes, broker
             // forcefully wants to close managed-ledger without waiting all resources to be closed.
-            if (!isClosingOrDeleting || closeWithoutWaitingClientDisconnect) {
-                fenceTopicToCloseOrDelete();
+            if (isClosingOrDeleting){
+                if (closeWithoutWaitingClientDisconnect){
+                    return this.fullyCloseFuture;
+                } else {
+                    // Why not return this.fullyCloseFuture ?
+                    // Just keep the same implementation as before.
+                    log.warn("[{}] Topic is already being closed or deleted", topic);
+                    return CompletableFuture.failedFuture(new TopicFencedException("Topic is already fenced"));
+                }
             } else {
-                log.warn("[{}] Topic is already being closed or deleted", topic);
-                closeFuture.completeExceptionally(new TopicFencedException("Topic is already fenced"));
-                return closeFuture;
+                fenceTopicToCloseOrDelete();
+                this.fullyCloseFuture = new CompletableFuture<>();
             }
         } finally {
             lock.writeLock().unlock();
         }
+        // Declare result.
+        CompletableFuture<Void> resultFuture = new CompletableFuture<>();
+        // Close limiters.
+        try {
+            closeLimiters();
+        } catch (Exception t){
```

Comment:
why should catch the Exception?

Reply:
Just to keep the logic the same as before: `close limiters` failure does not affect `topic close`.